### PR TITLE
DRS Localizer retries [BT-223]

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala [hunk 1]
```diff
@@ -42,9 +61,36 @@ class DrsLocalizerMain(drsUrl: String,
...
+                                    backoff: Option[CloudNioBackoff],
+                                    attempt: Int = 0): IO[DownloadResult] = {
+
+    def maybeResolveAndDownloadWithRetry(t: Throwable): IO[DownloadResult] = {
+      if (attempt < retries) {
+        backoff foreach { b => Thread.sleep(b.backoffMillis) }
+        logger.warn(s"Attempting retry $attempt of $retries retries to download $drsUrl", t)
+        resolveAndDownloadWithRetries(retries, downloaderBuilder, backoff map { _.next }, attempt + 1)
+      } else {
+        IO.raiseError(new Throwable(s"Exhausted $retries retries to resolve and download $drsUrl", t))
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:74]: Replace `Throwable` with `RuntimeException`

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala [hunk 2]
```diff
@@ -11,19 +11,54 @@ case class AccessUrlDownloader(accessUrl: AccessUrl, downloadLoc: String) extend
 
   def generateDownloadScript(): String = {
     val signedUrl = accessUrl.url
-    // TODO headers, refinements to retry strategy for certain 4xx statuses
-    s"""mkdir -p $$(dirname '$downloadLoc') && curl --location --retry 3 --retry-connrefused --retry-delay 10 --fail --output '$downloadLoc' '$signedUrl'"""
+    // TODO headers
+    s"""mkdir -p $$(dirname '$downloadLoc') && rm -f '$downloadLoc' && curl --silent --write-out '%{http_code}' --location --fail --output '$downloadLoc' '$signedUrl'"""
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:15]: Oof. Someday resumable downloads... someday.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala [hunk 3]
```diff
@@ -2,8 +2,16 @@ package drs.localizer.downloaders
...
-  def download: IO[ExitCode]
+sealed trait DownloadResult {
+  def exitCode: ExitCode
+}
+case object DownloadSuccess extends DownloadResult {
+  override def exitCode: ExitCode = ExitCode(0)
+}
+sealed trait DownloadFailure extends DownloadResult
+case class RetryableDownloadFailure(override val exitCode: ExitCode = ExitCode(0)) extends DownloadFailure
+case class NonRetryableDownloadFailure(override val exitCode: ExitCode = ExitCode(0)) extends DownloadFailure
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala:13]: Curiosity: why have defaults here? Do failures with `ExitCode(0)` have an implicit meaning?

**Review by kshakir — APPROVED** [review]
Nothing major 👍

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala (hunk 3 above)
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala]: I think this is leftover from an earlier version of the design, certainly not needed now.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala [hunk 4]
```diff
@@ -11,19 +11,54 @@ case class AccessUrlDownloader(accessUrl: AccessUrl, downloadLoc: String) extend
...
+  }
 
-    IO(ExitCode(returnCode))
+  def result(returnCode: Int, httpStatus: String, maskedUrl: String): DownloadResult = {
+    (returnCode, httpStatus) match {
+      case (i, _) if i == 7 => // ECONNREFUSED
+        RetryableDownloadFailure(exitCode = ExitCode(i))
+      case (i, s) =>
+        try {
+          val intStatus = Integer.parseInt(s)
```
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:48]: TOL: This is fine, though I wonder if there's some HTTP status class that we can use. I'd imagine such a thing would be able to say what class of result it is (i.e. 2xx or 5xx) as well as check for specific codes.

**Review by breilly2 — APPROVED** [review]
👍   Looks good to me.

TOL: I'm having trouble putting my finger on it exactly, but I feel like there may be some test coverage overlap. Maybe it's around having to manually stub `DownloaderBuilder` and the `Downloader`s it returns. Maybe I've just gotten so used to the less formal nature of JS. 🤷

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala (hunk 4 above)
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:48]: I found [this](https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.Status.Family.html#familyOf-int-) in JEE 7 and [this](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/http/HttpStatusFamily.html#of-int-) in the AWS SDK and [this](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpStatus.Series.html#resolve-int-) in Spring but none of those are available here.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala [hunk 5]
```diff
@@ -40,9 +59,36 @@ class DrsLocalizerMain(drsUrl: String,
...
+      if (attempt < retries) {
+        backoff foreach { b => Thread.sleep(b.backoffMillis) }
+        logger.warn(s"Attempting retry $attempt of $retries retries to download $drsUrl", t)
+        resolveAndDownloadWithRetries(retries, downloaderBuilder, backoff map { _.next }, attempt + 1)
+      } else {
+        IO.raiseError(new RuntimeException(s"Exhausted $retries retries to resolve and download $drsUrl", t))
+      }
+    }
+
+    resolveAndDownload(downloaderBuilder).redeemWith({
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:76]: Yes this is a little strange looking at first, but it's a useful pattern you'll see in a few places in the Cromwell code base. I was looking at Centaur before picking up this ticket so I was inspired by a [similar block](https://github.com/broadinstitute/cromwell/blob/develop/centaur/src/it/scala/centaur/AbstractCentaurTestCaseSpec.scala#L191) used for retries of failing tests.

**Pull request merged**